### PR TITLE
[ALLUXIO-2089] Pagination support for large directories in Swift

### PR DESCRIPTION
**Pull request by madanadit**
A more efficient listRecursive call is also added.

JIRA ticket: https://alluxio.atlassian.net/browse/ALLUXIO-2089

### CODE REVIEW

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 1]
```diff
@@ -552,17 +535,63 @@ private boolean isDirectory(String path) throws IOException {
...
+  /**
+   * Lists the files or folders which match the given prefix using pagination.
    *
    * @param prefix the prefix to match
+   * @param limit use delimiter
    * @return a collection of the files or folders matching the prefix, or null if not found
    * @throws IOException if path is not accessible, e.g. network issues
    */
-  private Collection<DirectoryOrObject> listInternal(final String prefix) throws IOException {
+  private Collection<DirectoryOrObject> listInternal(final String prefix, boolean limit)
```
Comment (calvinjia) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: Would it make sense to change `limit` -> `recursive` and write the code a bit differently? `limit` is just `!recursive` right?

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 2]
```diff
@@ -552,17 +535,63 @@ private boolean isDirectory(String path) throws IOException {
...
+      String child = stripFolderSuffixIfPresent(object.getName());
+      String noPrefix = CommonUtils.stripPrefixIfPresent(child, prefix);
+      if (!noPrefix.equals(self)) {
+        children.add(noPrefix);
+      } else {
+        foundSelf = true;
+      }
+    }
+
+    if (!foundSelf) {
```
Comment (calvinjia) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: Could the logic here be simplified by not checking each element and then using `children.contains(self)` as the condition?

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 3]
```diff
@@ -552,17 +535,63 @@ private boolean isDirectory(String path) throws IOException {
...
-   * Lists the files or folders which match the given prefix.
+   * Lists the files or folders in the path and remove the path itself.
+   *
+   * @param path the prefix to match
+   * @param recursive whether to do a recursive listing
+   * @return a collection of the files or folders matching the prefix, or null if not found
+   * @throws IOException if path is not accessible, e.g. network issues
+   */
+  private String[] listHelper(String path, boolean recursive) throws IOException {
+    String prefix = addFolderSuffixIfNotPresent(stripContainerPrefixIfPresent(path));
```
Comment (calvinjia) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: Given the meaning of `FolderSuffix` across under storages, this is a bit confusing. This line is really trying to accomplish `addPathSeparatorIfNotPresent` right?

**Comment by calvinjia** [comment]
@madanadit Two minor comments, also is it possible to have a test for this in the integration tests? Does not need to be in this PR as it should work for all the under storages.

**Comment by madanadit** [comment]
@calvinjia Thanks for the comments.

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java [hunk 4]
```diff
@@ -552,17 +535,63 @@ private boolean isDirectory(String path) throws IOException {
   }
 
   /**
-   * Lists the files or folders which match the given prefix.
+   * Lists the files or folders in the given path, not including the path itself.
+   *
+   * @param path the folder path whose children are listed
```
Comment (yupeng9) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: the path can also navigates to a file, right?

Comment (madanadit) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: No. The UnderFileSystem specifies that the path must be a directory. If path does not exist as directory, return null.

Comment (yupeng9) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: gotcha, then you can clarify this in the `@return` doc?

**Comment by calvinjia** [comment]
LGTM

File: underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java (hunk 4 above)
Comment (madanadit) [inline underfs/swift/src/main/java/alluxio/underfs/swift/SwiftUnderFileSystem.java]: done

**Comment by yupeng9** [comment]
LGTM

**Pull request merged**